proposals/hips/hip-1195.md

```diff
@@ -0,0 +1,348 @@
+# HIP-1195: Hiero hooks and an application to allowances
+
+HIP: https://hips.hedera.com/hip/hip-1195
+
+This HIP introduces **hooks**, programmable Hiero extension points that let users customize the behavior of their entities. The initial implementation focuses on **EVM hooks** and **account allowance hooks** as the first extension point.
+
+## New APIs
+
+#### HookExtensionPoint
+```
+enum HookExtensionPoint {
+    ACCOUNT_ALLOWANCE_HOOK = 0
+}
+```
+
+#### HookCreationDetails
+```
+// The details of a hook's creation.
+type HookCreationDetails {
+    // The extension point for the hook.
+    HookExtensionPoint extensionPoint
+
+    // The id to create the hook at.
+    uint64 hookId
+
+    // The hook implementation.
+    // A hook programmed in EVM bytecode that may access state or interact with external contracts.
+    LambdaEvmHook lambdaHook 
+
+    // A key that that can be used to remove or replace the hook.
+    Key adminKey
+}
+```
+
+#### LambdaEvmHook
+```
+// Definition of a lambda EVM hook.
+type LambdaEvmHook extends EvmHookSpec {
+    // Initial storage updates for the lambda, if any.
+    List<LambdaStorageUpdate> storageUpdates
+}
+
+// Shared specifications for an EVM hook. May be used for any extension point.
+type EvmHookSpec {
+    // The source of the EVM bytecode for the hook.
+    ContractId contractId
+}
+```
+
+#### LambdaStorageUpdate
+```
+// Specifies a key/value pair in the storage of a lambda, either by the explicit storage slot contents; or by a combination of a Solidity mapping's slot key and the key into that mapping.
+type LambdaStorageUpdate {
+    // oneof (mutually exclusive):
+
+    // An explicit storage slot update.
+    LambdaStorageSlot storageSlot  
+
+    // An implicit storage slot update specified as a Solidity mapping entry.
+    LambdaMappingEntry mappingEntry 
+}
+
+type LambdaStorageSlot {
+    // 32-byte storage slot key
+    bytes key
+
+    // 32-byte storage slot value (empty to delete)
+    bytes value
+}
+
+type LambdaMappingEntry {
+    // Slot corresponding to Solidity mapping
+    bytes mappingSlot
+
+    // 32-byte key of mapping entry
+    bytes key
+
+    // 32-byte value of mapping entry (empty to delete)
+    bytes value
+}
+```
+
+#### HookCall and EvmHookCall
+```
+// Specifies a call to a hook from within a transaction where the hook owner is implied by the point of use.
+type HookCall {
+    // The id of the hook to call.
+    uint64 hookId
+
+    // Specification of how to call an EVM hook.
+    EvmHookCall evmHookCall 
+}
+
+// Specifies details of a call to an EVM hook.
+type EvmHookCall {
+    // Call data to pass to the hook.
+    bytes data
+
+    // The gas limit to use.
+    uint64 gasLimit
+}
+```
+
+#### LambdaSStoreTransaction
+```
+// Adds or removes key/value pairs in the storage of a lambda.
+type LambdaSStoreTransaction {
+    // The id of the lambda whose storage is being updated.
+    CreatedHookId hookId
```

Comment:
What is this `CreatedHookId` type? Is it defined somewhere and I'm just missing it?